### PR TITLE
xSPUserProfileSyncConn

### PR DESCRIPTION
**Pull request by camiloborges**
AD import not implemented yet.

### CODE REVIEW

**Comment by BrianFarnhill** [comment]
@camiloborges Failing unit tests there brother. Also when you say AD import not yet supported, I'm assuming you mean we do work with AD that is using the full user profile sync service but not yet the basic LDAP option?

**Comment by camiloborges** [comment]
Hello.

Will look into the unit tests, it was missing one new method inside .util. Obviously I screwed something up as I thought I had already committed and pushed the fix.

Ad import is the other ldap way to do ups sync, I use that in clients with adfs.



Camilo Borges
Mob:0488 056 964
Sent via iPhone while out and about

**Comment by camiloborges** [comment]
Reference:
http://blogs.technet.com/b/meacoex/archive/2013/08/04/step-by-step-active-directory-import-for-sharepoint-2013.aspx

File: Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1 [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    $params = $args[0]
+        
+
+    if ($params.ContainsKey("InstallAccount")) { $params.Remove("InstallAccount") | Out-Null }
+    $ups = Get-SPServiceApplication -Name $params.UserProfileService -ErrorAction SilentlyContinue 
+                
+    if ($null -eq $ups) { 
+        throw "User Profile Service Application $($params.UserProfileService) not found"
+    }
+    $context = Get-xSharePointServiceContext  $ups.ServiceApplicationProxyGroup
```
Comment (camiloborges) [inline Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1]: @BrianFarnhill hey, I think Get-xSharePointServiceContext should be replaced by Get-SPServiceContext, but I had some issues while I battled with the API to get the connection working. 
I will test it again with Get-SPServiceContext when I get to do the other bits

File: Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1 [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            }
+            $namingContext = $connection.NamingContexts | select -first 1
+            if($namingContext -eq $null){
+                return $null
+            }
+            $accountCredentials = "$($connection.AccountDomain)\$($connection.AccountUsername)"
+            $domainController = $namingContext.PreferredDomainControllers | select -First 1
+            return @{
+                        UserProfileService = $UserProfileService
+                        Forest = $connection.Server #"contoso.com" #TODO: GetCorrect Forest
```
Comment (BrianFarnhill) [inline Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1]: Is this "todo" still outstanding mate?

File: Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1 (hunk 1 above)
Comment (BrianFarnhill) [inline Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1]: Did you get to your testing on this? Whats the issue you are seeing here?

File: Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1 [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        [parameter(Mandatory = $true)]  [System.String] $Name,
+        [parameter(Mandatory = $true)]  [System.String] $Forest,
+        [parameter(Mandatory = $true)]  [System.Management.Automation.PSCredential] $ConnectionCredentials,
+        [parameter(Mandatory = $true)] [System.String] $UserProfileService,
+        [parameter(Mandatory = $true)] [System.String[]] $IncludedOUs,
+        [parameter(Mandatory = $false)] [System.String[]] $ExcludedOUs,
+        [parameter(Mandatory = $false)] [System.String] $Server,
+        [parameter(Mandatory = $false)] [System.String] $Force,
+        [parameter(Mandatory = $false)] [System.Boolean] $UseSSL,
+        [parameter(Mandatory = $false)] [System.String] $ConnectionType,
```
Comment (BrianFarnhill) [inline Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1]: Looks like you specify the ValueMap in the schema for this, so you need the correspinding value map here as well. The recently fixed MOF schema tests will pick this up on the next test run too.

**Comment by BrianFarnhill** [comment]
@camiloborges Just a few things to review in here for you mate, also need the "unreleased" section of the readme to be updated as well

File: Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1 (hunk 1 above)
Comment (camiloborges) [inline Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSyncConnection/MSFT_xSPUserProfileSyncConnection.psm1]: So, my problem was that to get the Context we need to have a url of an application. So for the ups property I got the CA url. I think this resource could/should use the same approach so I will update it before tuesday. shouldn't change much but I will have to update the unit tests.

File: Modules/xSharePoint/Modules/xSharePoint.Util/xSharePoint.Util.psm1 [hunk 4]
```diff
@@ -27,6 +27,17 @@ function Get-xSharePointAssemblyVersion() {
     return (Get-Command $PathToAssembly).FileVersionInfo.FileMajorPart
 }
 
+function Get-xSharePointServiceContext {
+    [CmdletBinding()]
+    param
+    (
+        [parameter(Mandatory = $true,Position=1)]
+        $proxyGroup
```
Comment (TravisEz13) [inline Modules/xSharePoint/Modules/xSharePoint.Util/xSharePoint.Util.psm1]: Function parameter names should be Pascal case per [style guidance](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md#parameter-names-should-use-pascalcase-)

**Comment by camiloborges** [comment]
hey @BrianFarnhill  check if this is now ok.

**Comment by TravisEz13** [comment]
Thanks for fixing my notes.

**Pull request merged**